src/main/java/pl/allegro/tech/discovery/consul/recipes/watch/ConsulLongPollCallback.java

```diff
@@ -185,21 +192,23 @@ private boolean contentChanged(byte[] newContent) {
         return !Arrays.equals(oldContent, newContent);
     }
 
-    private long updateIndex(Response response) {
-        String indexString = response.header("X-Consul-Index");
-        if (indexString == null) {
-            logger.error("There was no X-Consul-Index header in response for {} endpoint, retrying", endpoint);
-            return -1;
-        }
-
-        long index = Long.valueOf(indexString);
-        if (currentIndex.get() == index) {
-            stats.indexNotChanged();
-            logger.trace("Discarding event on endpoint {} index {} as index did not change", endpoint, index);
-            return -1;
-        } else {
-            currentIndex.set(index);
-            return index;
+    private void handleUpdate(long newIndex, ResponseBody body) throws IOException {
+        // update index and process content
+        long lastIndex = currentIndex.get();
+
+        if (newIndex == lastIndex) {
+            handleIndexNotChanged(newIndex);
+        } else if (newIndex < lastIndex) { // index backwards -> reset
+            currentIndex.set(0);
+            handleIndexBackwards(lastIndex, newIndex);
+        } else { // index moving forward
+            currentIndex.set(newIndex);
```

Comment:
I'd rather extract whole else to a method (i.e. handleIndexUpdateBasedOnContent) to match above